tools/daemon/flow.c

```diff
@@ -170,24 +171,55 @@ int add_flow(struct store_pid *pid_value, struct store_flow *value)
 		}
 	}
 	if (cur_entry == &pid_value->flow_list) {
+		struct ifaddrs *ifaddr, *ifa;
+		char *addr = NULL;
+		int handle = 1;
+
 		/* This cleanup is done just to support verification */
 		sys_exec("tc qdisc del dev %s handle ffff: ingress > /dev/null 2>&1 || echo $?", tap_name);
 
-		/* Create filter to redirect traffic from tap device to netvsc device */
+		/* Create rules to process ingress trafic on tap device */
 		out_buf = sys_exec("tc qdisc add dev %s handle ffff: ingress > /dev/null 2>&1 || echo $?", tap_name);
 		if (NULL == out_buf || (out_buf[0] != '\0' && out_buf[0] != '0')) {
 			log_error("[%d] failed tc qdisc add dev %s output: %s\n",
 					pid, tap_name, (out_buf ? out_buf : "n/a"));
-			free(cur_element);
 			rc = -EFAULT;
 			goto err;
 		}
-		out_buf = sys_exec("tc filter add dev %s parent ffff: protocol all u32 match u8 0 0 action mirred egress redirect dev %s > /dev/null 2>&1 || echo $?",
-				tap_name, if_name);
+
+		if (!getifaddrs(&ifaddr)) {
+			for (ifa = ifaddr; NULL != ifa; ifa = ifa->ifa_next) {
+				if (ifa->ifa_addr->sa_family == AF_INET &&
+						value->if_id == if_nametoindex(ifa->ifa_name)) {
+					addr = inet_ntoa(((struct sockaddr_in *)ifa->ifa_addr)->sin_addr);
+
+					/* Create filter to redirect traffic from tap device to lo device in case destination address relates netvsc */
+					out_buf = sys_exec("tc filter add dev %s protocol ip parent ffff: prio 1 "
+								"handle ::%d u32 ht 800:: "
+								"match ip dst %s/32 action mirred egress redirect dev lo "
+								"> /dev/null 2>&1 || echo $?",
+								tap_name, handle, addr);
+					if (NULL == out_buf || (out_buf[0] != '\0' && out_buf[0] != '0')) {
+						log_error("[%d] failed tc filter add dev %s redirect to lo output: %s\n",
+								pid, tap_name, (out_buf ? out_buf : "n/a"));
+						rc = -EFAULT;
+						goto err;
+					}
+					handle++;
+				}
+			}
+			freeifaddrs(ifaddr);
+		}
+
+		/* Create filter to redirect traffic from tap device to netvsc device */
+		out_buf = sys_exec("tc filter add dev %s protocol ip parent ffff: prio 1 "
+					"handle ::%d u32 ht 800:: "
+					"match u8 0 0 action mirred egress redirect dev %s "
+					"> /dev/null 2>&1 || echo $?",
+					tap_name, handle, if_name);
 		if (NULL == out_buf || (out_buf[0] != '\0' && out_buf[0] != '0')) {
-			log_error("[%d] failed tc filter add dev %s output: %s\n",
-					pid, tap_name, (out_buf ? out_buf : "n/a"));
-			free(cur_element);
+			log_error("[%d] failed tc filter add dev %s redirect to %s output: %s\n",
+					pid, tap_name, if_name, (out_buf ? out_buf : "n/a"));
```

Comment:
why not sending a list of ip's from VMA instead of  this search ?
we need to use diffrent prio -
all local ip's - prio x
all the rest- prio < x.

The other approch is to set the rule based on the dst_entry value....